### DISCUSSION TITLE
JS Engine support

### DISCUSSION THREAD

**Discussion by amsam0** [chrisgurney/obsidian-note-toolbar#93] [Ideas]
Hi, this plugin looks really useful and I would love to start using it in my vault as a replacement for some Meta Bind buttons I use. However, most of these buttons are powered by custom javascript code run through the JS Engine plugin. While the note toolbar plugin supports running commands, JS Engine's "run JavaScript file" command requires you to pick a file when you run it, making it suboptimal for usage with Note Toolbar. It would be really nice if Note Toolbar had a item type that runs a JavaScript file.

~~As for actual implementation details, I'm not sure if JS Engine has a proper API but I know Meta Bind uses it so that may be worth taking a look at. (After a quick look at JS Engine's source code, it seems like the plugin's `api` property has an "internal" API class that is used for executing files)~~ Here's a link to how Meta Bind does it: https://github.com/mProjectsCode/obsidian-meta-bind-plugin/blob/0151349165be77a8e70a8f6f969053baceb366f4/packages/obsidian/src/ObsidianInternalAPI.ts#L79-L106

**Comment by chrisgurney** [comment]
@naturecodevoid Thanks for the suggestion! This is something I've wanted to look into, so I really appreciate the link to see how they support it.

Ideally I wouldn't want to rely on another plugin, but I'm guessing [JS Engine](https://github.com/mProjectsCode/obsidian-js-engine-plugin) and the more popular [Custom JS](https://github.com/saml-dev/obsidian-custom-js) does a bunch of stuff under the covers that's needed that I'm not thinking of. What's the difference between these two? I'd probably aim to support the more popular of the two first.

**Reply by amsam0** [reply]
I had never heard about the Custom JS plugin until you linked it, so I don't know how it differs from JS Engine. However, I'd imagine that each plugin would take about 10 lines of code to use their API, so I think it would be pretty reasonable to support both. Unless you have any objections, I'll probably try to get in a PR to support them.

**Comment by chrisgurney** [comment]
See also [this discussion](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/110) about scripting and an API.

**Comment by chrisgurney** [comment]
Support for JS Engine, Dataview, and Templater is now available in [the beta](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/171)!

**Reply by lumiarna** [reply]
Hi, are there any plans to support Datacore, the new product from the Dataview author?
https://github.com/blacksmithgu/datacore

**Reply by chrisgurney** [reply]
Hi @lumiarna I've just created [this discussion](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/462) for people to weigh in on whether Note Toolbar should support Datacore. Can you please vote on it, and add any thoughts you might have? Thanks!

**Comment by chrisgurney** [comment]
@amsam0 Support for JS Engine is now in the latest release.

**Reply by amsam0** [reply]
Thank you!! I've been trying the beta and it's been working great.

My only request: it would be great if we could set the label with JS engine, so we don't have to rely on templater. With templater, there is also a split second where the template is shown instead of the output.

**Reply by chrisgurney** [reply]
> With templater, there is also a split second where the template is shown instead of the output.

@amsam0 Can you share more details about this one? How are you using Templater here? Curious about that split-second thing.

**Reply by amsam0** [reply]
I was testing it out with `<% tp.file.name %>` as the label. When I switched between daily notes, the label showed the template for a small amount of time before switching to the result of the template.

**Reply by chrisgurney** [reply]
> I was testing it out with `<% tp.file.name %>` as the label. When I switched between daily notes, the label showed the template for a small amount of time before switching to the result of the template.

Ahh, it sounds like you're also experiencing what's described in #198. Thanks for the details.

**Comment by chrisgurney** [comment]
@amsam0 I've been thinking about your suggestion:

> It would be great if we could set the label with JS engine, so we don't have to rely on templater.

A couple thoughts:

- What labels would you like to see dynamically updated? (i.e., What are you using Templater for to do this?)
  - Given the above, how can JS Engine get this info (file name? date?)? One example of a Templater command and corresponding JavaScript line would be great.
- Might need a setting specific to JS Engine to define the special character that tells Note Toolbar that it's a JS expression that needs evaluating. (For Templater that's `<%` and Dataview uses `=`.)

**Reply by amsam0** [reply]
I was using templater to display the name of my file. Here's an example of JS Engine doing that:

````js
```js-engine
return engine.markdown.create(app.workspace.activeEditor.file.basename)
```
````

which should be equivalent to `<% tp.file.title %>`.

**Reply by chrisgurney** [reply]
@amsam0 Perfect, thank you.

**Reply by chrisgurney** [reply]
I've moved this idea over to #211

**Reply by laktiv** [reply]
There is one more - [CodeScript Toolkit](https://github.com/mnaoumov/obsidian-codescript-toolkit). If I understood right, he combines a lot of possibilities. this mf could run everything. Didn't try a lot, but what tried, he provided result. And for me a plus in-md console.

**Reply by chrisgurney** [reply]
Thanks @laktiv , I'm aware of that one but haven't had any requests for it, nor have I looked into whether I can write an integration to it. 

If there is demand, I can take a look, for sure.

**Reply by laktiv** [reply]
> Thanks @laktiv , I'm aware of that one but haven't had any requests for it, nor have I looked into whether I can write an integration to it.
> 
> If there is demand, I can take a look, for sure.

No-no-no. Not request for sure. Just told as example, it kinda universal for every plugin.

but, a few days ago, somewhere here I saw Datacore. This is for sure very interesting
